flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/filesystem/DeserializationSchemaAdapter.java

```diff
@@ -0,0 +1,316 @@
+	private final RowType projectedRowType;
+
+	private final List<String> partitionKeys;
+	private final String defaultPartValue;
+
+	private final int[] toProjectedField;
+	private final RowData.FieldGetter[] formatFieldGetters;
+
+	public DeserializationSchemaAdapter(
+			DeserializationSchema<RowData> deserializationSchema,
+			TableSchema schema,
+			int[] projectFields,
+			List<String> partitionKeys,
+			String defaultPartValue) {
+		this.deserializationSchema = deserializationSchema;
+		this.fieldNames = schema.getFieldNames();
+		this.fieldTypes = schema.getFieldDataTypes();
+		this.projectFields = projectFields;
+		this.partitionKeys = partitionKeys;
+		this.defaultPartValue = defaultPartValue;
+
+		List<String> projectedNames = Arrays.stream(projectFields)
+				.mapToObj(idx -> schema.getFieldNames()[idx])
+				.collect(Collectors.toList());
+
+		this.projectedRowType = RowType.of(
+				Arrays.stream(projectFields).mapToObj(idx ->
+						schema.getFieldDataTypes()[idx].getLogicalType()).toArray(LogicalType[]::new),
+				projectedNames.toArray(new String[0]));
+
+		List<String> formatFields = Arrays.stream(schema.getFieldNames())
+				.filter(field -> !partitionKeys.contains(field))
+				.collect(Collectors.toList());
+
+		List<String> formatProjectedFields = projectedNames.stream()
+				.filter(field -> !partitionKeys.contains(field))
+				.collect(Collectors.toList());
+
+		this.toProjectedField = formatProjectedFields.stream()
+				.mapToInt(projectedNames::indexOf)
+				.toArray();
+
+		this.formatFieldGetters = new RowData.FieldGetter[formatProjectedFields.size()];
+		for (int i = 0; i < formatProjectedFields.size(); i++) {
+			String name = formatProjectedFields.get(i);
+			this.formatFieldGetters[i] = RowData.createFieldGetter(
+					schema.getFieldDataType(name).get().getLogicalType(),
+					formatFields.indexOf(name));
+		}
+	}
+
+	private DeserializationSchema<RowData> createDeserialization() throws IOException {
+		try {
+			DeserializationSchema<RowData> deserialization = InstantiationUtil.clone(deserializationSchema);
+			deserialization.open(new DeserializationSchema.InitializationContext() {
+				@Override
+				public MetricGroup getMetricGroup() {
+					throw new UnsupportedOperationException("MetricGroup is unsupported in BulkFormat.");
+				}
+
+				@Override
+				public UserCodeClassLoader getUserCodeClassLoader() {
+					return (UserCodeClassLoader) Thread.currentThread().getContextClassLoader();
+				}
+			});
+			return deserialization;
+		} catch (Exception e) {
+			throw new IOException(e);
+		}
+	}
+
+	@Override
+	public Reader createReader(Configuration config, FileSourceSplit split) throws IOException {
+		return new Reader(config, split);
+	}
+
+	@Override
+	public Reader restoreReader(Configuration config, FileSourceSplit split) throws IOException {
+		Reader reader = new Reader(config, split);
+		reader.seek(split.getReaderPosition().get().getRecordsAfterOffset());
+		return null;
```

Comment:
return reader?

Reply:
Streaming restore case: no test cover

Reply:
But there is no way, because the current file system does not support streaming reading